### PR TITLE
Add production build

### PR DESCRIPTION
**Pull request by ludohenin**
* Bundles the app and all dependencies (except angular2/RxJS itself)
* Bundling and lazy loading ready

__Edit:__

Some optimization is still needed.

* [ ] tsc: emit js without helpers (it's a lot of duplicated code). 
* [ ] Provide global typescript helpers.

### CODE REVIEW

File: tools/config.ts [hunk 1]
```diff
@@ -17,50 +17,78 @@ export const APP_SRC              = 'app';
...
   defaultJSExtensions: true,
   paths: {
+    'bootstrap': `${APP_ROOT}bootstrap`,
     '*': `${APP_BASE}node_modules/*`
   }
 };
 
+const SYSTEM_CONFIG_PROD = {
+  defaultJSExtensions: true,
+  bundles: {
```
Comment (mgechev) [inline tools/config.ts]: Love this!

File: tools/tasks/build.index.ts [hunk 2]
```diff
@@ -22,8 +22,13 @@ export = function buildIndexDev(gulp, plugins) {
...
-    return DEV_DEPENDENCIES.concat(APP_ASSETS)
+    return DEPENDENCIES.concat(APP_ASSETS)
       .filter(dep => dep['inject'] && dep['inject'] === (name || true))
-      .map(dep => dep.src);
+      .map(mapPath);
+  }
+
+  function mapPath(dep) {
+    let prodPath = `${dep.dest}/${dep.src.split('/').pop()}`;
+    return ('prod' === ENV ? prodPath : dep.src );
```
Comment (mgechev) [inline tools/tasks/build.index.ts]: We can have an enum:

```ts
enum BUILD_TARGET {
  PRODUCTION,
  DEVELOPMENT
};
```
or something like that.

File: tools/config.ts [hunk 3]
```diff
@@ -17,50 +17,78 @@ export const APP_SRC              = 'app';
...
 export const VERSION              = appVersion();
 
 export const VERSION_NPM          = '3.0.0';
 export const VERSION_NODE         = '4.0.0';
 
 // Declare NPM dependencies (Note that globs should not be injected).
-export const DEV_DEPENDENCIES = [
-  { src: 'systemjs/dist/system-polyfills.js' },
+export const NPM_DEPENDENCIES = [
+  { src: 'systemjs/dist/system-polyfills.js', dest: LIB_DEST },
```
Comment (mgechev) [inline tools/config.ts]: Can't we get rid of `LIB_DEST`? We may declare it in production config, something like:

```ts
const PROD_CONFIG = {
  CSS_DEST,
  LIB_DEST
};
```

**Comment by NathanWalker** [comment]
This is really excellent. Much applause! :)

**Pull request merged**